File: js/src/lib/execution-resume-all.js
/**
 * `--resume-all`: re-attach or reconcile every execution still marked running.
 *
 * When the supervisor host restarts, the detached docker completion watcher
 * (`docker logs -f ... ; docker inspect ... ; footer`) dies with it. The
 * container keeps running but nothing streams its output into the session log
 * any more, and nothing will ever write the exit footer — the record stays
 * `executing` forever. `--resume-all` repairs that state:
 *
 * - reattached: a live docker container gets a fresh completion watcher.
 * - running:    a live screen/tmux session needs nothing (its logging is
 *               in-session), it is only reported.
 * - reconciled: the session is gone, so the record is finalized from the same
 *               evidence `--status` uses (docker exit code, log footer).
 * - unknown:    the backend cannot be probed locally (ssh), left untouched.
 *
 * Commands are never silently restarted here: resuming actual work is an
 * explicit, per-session decision made with `--resume`.
 */

const {
  getDockerContainerCleanupPolicy,
  startDetachedDockerCompletionWatcher,
} = require('./docker-cleanup');
const { escapeForLinksNotation } = require('./output-blocks');
const { runCommand } = require('./execution-control');
const { SessionState, probeSession } = require('./session-probe');

/**
 * Outcomes reported for each execution that was still marked running.
 */
const ResumeAllAction = {
  REATTACHED: 'reattached',
  RUNNING: 'running',
  RECONCILED: 'reconciled',
  UNKNOWN: 'unknown',
};

function formatResumeAllAsLinksNotation(summary) {
  const lines = ['executionResumeAll', `  count ${summary.count}`];
  for (const entry of summary.executions) {
    lines.push('  execution');
    lines.push(`    uuid ${escapeForLinksNotation(entry.uuid)}`);
    lines.push(`    backend ${escapeForLinksNotation(entry.backend)}`);
    lines.push(`    sessionName ${escapeForLinksNotation(entry.sessionName)}`);
    lines.push(`    state ${escapeForLinksNotation(entry.state)}`);
    lines.push(`    action ${escapeForLinksNotation(entry.action)}`);
    if (entry.exitCode !== null && entry.exitCode !== undefined) {
      lines.push(`    exitCode ${entry.exitCode}`);
    }
    lines.push(`    message ${escapeForLinksNotation(entry.message)}`);
  }
  return lines.join('\n');
}

function formatResumeAllAsText(summary) {
  if (summary.count === 0) {
    return 'No executions are currently marked as running.';
  }
  const lines = [`Executions still marked running: ${summary.count}`, ''];
  for (const entry of summary.executions) {
    lines.push(`${entry.action.toUpperCase()}  ${entry.uuid}`);
    lines.push(`  Backend:      ${entry.backend}`);
    lines.push(`  Session Name: ${entry.sessionName}`);
    lines.push(`  State:        ${entry.state}`);
    lines.push(`  ${entry.message}`);
    lines.push('');
  }
  return lines.join('\n').trimEnd();
}

function formatResumeAll(summary, outputFormat) {
  if (outputFormat === 'json') {
    return JSON.stringify(summary, null, 2);
  }
  if (outputFormat === 'text') {
    return formatResumeAllAsText(summary);
  }
  return formatResumeAllAsLinksNotation(summary);
}

function describeRecord(record, probe, action, message, exitCode = null) {
  return {
    uuid: record.uuid,
    backend:
      probe.backend || (record.options && record.options.isolated) || null,
    sessionName:
      probe.sessionName ||
      (record.options && record.options.sessionName) ||
      null,
    state: probe.state,
    action,
    exitCode,
    message,
  };
}

/**
 * Re-attach or reconcile every execution still marked running.
 * @param {?object} store - Execution store
 * @param {object} deps - {outputFormat, probe, runner, startWatcher, reconcile}
 * @returns {{success: boolean, output?: string, error?: string}}
 */
function resumeAllExecutions(store, deps = {}) {
  if (!store) {
    return { success: false, error: 'Execution tracking is disabled.' };
  }

  const runner = deps.runner || runCommand;
  const probeFn = deps.probe || ((record) => probeSession(record, runner));
  const startWatcher =
    deps.startWatcher || startDetachedDockerCompletionWatcher;
  const reconcile =
    deps.reconcile || require('./status-formatter').enrichDetachedStatus;

  let records;
  try {
    records = store.getExecuting();
  } catch (err) {
    return { success: false, error: err.message };
  }

  const executions = [];

  for (const record of records) {
    const probe = probeFn(record);
    const sessionName = probe.sessionName;

    if (probe.alive && probe.backend === 'docker') {
      startWatcher(
        sessionName,
        getDockerContainerCleanupPolicy(record.options || {}),
        record.logPath || null
      );
      executions.push(
        describeRecord(
          record,
          probe,
          ResumeAllAction.REATTACHED,
          `Re-attached completion watcher to running container: ${sessionName}`
        )
      );
      continue;
    }

    if (probe.alive) {
      executions.push(
        describeRecord(
          record,
          probe,
          ResumeAllAction.RUNNING,
          `Session is still running: ${sessionName}`
        )
      );
      continue;
    }

    if (probe.state === SessionState.UNKNOWN) {
      executions.push(
        describeRecord(
          record,
          probe,
          ResumeAllAction.UNKNOWN,
          `Session liveness cannot be probed locally: ${sessionName || record.uuid}`
        )
      );
      continue;
    }

    const reconciled = reconcile(record);
    if (reconciled && reconciled.status === 'executed') {
      store.save(reconciled);
      executions.push(
        describeRecord(
          record,
          probe,
          ResumeAllAction.RECONCILED,
          `Session ended while unsupervised; record finalized. Use \`$ --resume ${record.uuid}\` to continue it.`,
          reconciled.exitCode
        )
      );
      continue;
    }

    executions.push(
      describeRecord(
        record,
        probe,
        ResumeAllAction.UNKNOWN,
        `Session is not running but no terminal result could be resolved: ${sessionName || record.uuid}`
      )
    );
  }

  return {
    success: true,
    output: formatResumeAll(
      { count: executions.length, executions },
      deps.outputFormat
    ),
  };
}

module.exports = {
  ResumeAllAction,
  formatResumeAll,
  formatResumeAllAsLinksNotation,
  formatResumeAllAsText,
  resumeAllExecutions,
};

File: rust/src/lib/execution_resume_all.rs
//! `--resume-all`: re-attach or reconcile every execution still marked running.
//!
//! When the supervisor host restarts, the detached docker completion watcher
//! (`docker logs -f ... ; docker inspect ... ; footer`) dies with it. The
//! container keeps running but nothing streams its output into the session log
//! any more, and nothing will ever write the exit footer — the record stays
//! `executing` forever. `--resume-all` repairs that state:
//!
//! - reattached: a live docker container gets a fresh completion watcher.
//! - running: a live screen/tmux session needs nothing (its logging is
//!   in-session), it is only reported.
//! - reconciled: the session is gone, so the record is finalized from the same
//!   evidence `--status` uses (docker exit code, log footer).
//! - unknown: the backend cannot be probed locally (ssh), left untouched.
//!
//! Commands are never silently restarted here: resuming actual work is an
//! explicit, per-session decision made with `--resume`.

use serde_json::json;

use crate::execution_control::{CommandRunner, SystemCommandRunner};
use crate::execution_resume::{ResumeHooks, SystemResumeHooks};
use crate::execution_store::{ExecutionRecord, ExecutionStatus, ExecutionStore};
use crate::output_blocks::escape_for_links_notation;
use crate::session_probe::{probe_session, SessionProbe, SessionState};

/// Outcomes reported for each execution that was still marked running.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ResumeAllAction {
    Reattached,
    Running,
    Reconciled,
    Unknown,
}

impl ResumeAllAction {
    pub fn as_str(self) -> &'static str {
        match self {
            ResumeAllAction::Reattached => "reattached",
            ResumeAllAction::Running => "running",
            ResumeAllAction::Reconciled => "reconciled",
            ResumeAllAction::Unknown => "unknown",
        }
    }
}

/// One line of the `--resume-all` report.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ResumeAllEntry {
    pub uuid: String,
    pub backend: Option<String>,
    pub session_name: Option<String>,
    pub state: SessionState,
    pub action: ResumeAllAction,
    pub exit_code: Option<i32>,
    pub message: String,
}

/// Result of `resume_all_executions`.
pub struct ExecutionResumeAllResult {
    pub success: bool,
    pub output: Option<String>,
    pub error: Option<String>,
}

fn describe_record(
    record: &ExecutionRecord,
    probe: &SessionProbe,
    action: ResumeAllAction,
    message: String,
    exit_code: Option<i32>,
) -> ResumeAllEntry {
    ResumeAllEntry {
        uuid: record.uuid.clone(),
        backend: probe.backend.clone().or_else(|| {
            record
                .options
                .get("isolated")
                .and_then(|value| value.as_str())
                .map(str::to_string)
        }),
        session_name: probe.session_name.clone().or_else(|| {
            record
                .options
                .get("sessionName")
                .and_then(|value| value.as_str())
                .map(str::to_string)
        }),
        state: probe.state,
        action,
        exit_code,
        message,
    }
}

fn format_as_links_notation(entries: &[ResumeAllEntry]) -> String {
    let mut lines = vec![
        "executionResumeAll".to_string(),
        format!("  count {}", entries.len()),
    ];
    for entry in entries {
        lines.push("  execution".to_string());
        lines.push(format!(
            "    uuid {}",
            escape_for_links_notation(&entry.uuid)
        ));
        lines.push(format!(
            "    backend {}",
            escape_for_links_notation(entry.backend.as_deref().unwrap_or(""))
        ));
        lines.push(format!(
            "    sessionName {}",
            escape_for_links_notation(entry.session_name.as_deref().unwrap_or(""))
        ));
        lines.push(format!(
            "    state {}",
            escape_for_links_notation(entry.state.as_str())
        ));
        lines.push(format!(
            "    action {}",
            escape_for_links_notation(entry.action.as_str())
        ));
        if let Some(exit_code) = entry.exit_code {
            lines.push(format!("    exitCode {}", exit_code));
        }
        lines.push(format!(
            "    message {}",
            escape_for_links_notation(&entry.message)
        ));
    }
    lines.join("\n")
}

fn format_as_text(entries: &[ResumeAllEntry]) -> String {
    if entries.is_empty() {
        return "No executions are currently marked as running.".to_string();
    }
    let mut lines = vec![
        format!("Executions still marked running: {}", entries.len()),
        String::new(),
    ];
    for entry in entries {
        lines.push(format!(
            "{}  {}",
            entry.action.as_str().to_uppercase(),
            entry.uuid
        ));
        lines.push(format!(
            "  Backend:      {}",
            entry.backend.as_deref().unwrap_or("")
        ));
        lines.push(format!(
            "  Session Name: {}",
            entry.session_name.as_deref().unwrap_or("")
        ));
        lines.push(format!("  State:        {}", entry.state.as_str()));
        lines.push(format!("  {}", entry.message));
        lines.push(String::new());
    }
    lines.join("\n").trim_end().to_string()
}

fn format_as_json(entries: &[ResumeAllEntry]) -> String {
    let executions: Vec<_> = entries
        .iter()
        .map(|entry| {
            json!({
                "uuid": entry.uuid,
                "backend": entry.backend,
                "sessionName": entry.session_name,
                "state": entry.state.as_str(),
                "action": entry.action.as_str(),
                "exitCode": entry.exit_code,
                "message": entry.message,
            })
        })
        .collect();
    serde_json::to_string_pretty(&json!({
        "count": entries.len(),
        "executions": executions,
    }))
    .unwrap_or_else(|_| "{}".to_string())
}

/// Format a `--resume-all` report in the requested output format.
pub fn format_resume_all(entries: &[ResumeAllEntry], output_format: Option<&str>) -> String {
    match output_format {
        Some("json") => format_as_json(entries),
        Some("text") => format_as_text(entries),
        _ => format_as_links_notation(entries),
    }
}

/// Re-attach or reconcile every execution still marked running.
pub fn resume_all_executions(
    store: Option<&ExecutionStore>,
    output_format: Option<&str>,
) -> ExecutionResumeAllResult {
    resume_all_executions_with(
        store,
        output_format,
        &SystemCommandRunner,
        &SystemResumeHooks,
    )
}

/// Re-attach or reconcile with an injectable command runner and hooks.
pub fn resume_all_executions_with<R: CommandRunner, H: ResumeHooks>(
    store: Option<&ExecutionStore>,
    output_format: Option<&str>,
    runner: &R,
    hooks: &H,
) -> ExecutionResumeAllResult {
    let Some(store) = store else {
        return ExecutionResumeAllResult {
            success: false,
            output: None,
            error: Some("Execution tracking is disabled.".to_string()),
        };
    };

    let mut entries: Vec<ResumeAllEntry> = Vec::new();

    for record in store.get_executing() {
        let probe = probe_session(&record, runner);
        let session_name = probe.session_name.clone().unwrap_or_default();

        if probe.alive && probe.backend.as_deref() == Some("docker") {
            hooks.start_watcher(&session_name, &record);
            entries.push(describe_record(
                &record,
                &probe,
                ResumeAllAction::Reattached,
                format!(
                    "Re-attached completion watcher to running container: {}",
                    session_name
                ),
                None,
            ));
            continue;
        }

        if probe.alive {
            entries.push(describe_record(
                &record,
                &probe,
                ResumeAllAction::Running,
                format!("Session is still running: {}", session_name),
                None,
            ));
            continue;
        }

        if probe.state == SessionState::Unknown {
            entries.push(describe_record(
                &record,
                &probe,
                ResumeAllAction::Unknown,
                format!(
                    "Session liveness cannot be probed locally: {}",
                    if session_name.is_empty() {
                        record.uuid.clone()
                    } else {
                        session_name.clone()
                    }
                ),
                None,
            ));
            continue;
        }

        let reconciled = hooks.reconcile(&record);
        if reconciled.status == ExecutionStatus::Executed {
            let exit_code = reconciled.exit_code;
            let _ = store.save(&reconciled);
            entries.push(describe_record(
                &record,
                &probe,
                ResumeAllAction::Reconciled,
                format!(
                    "Session ended while unsupervised; record finalized. Use `$ --resume {}` to continue it.",
                    record.uuid
                ),
                exit_code,
            ));
            continue;
        }

        entries.push(describe_record(
            &record,
            &probe,
            ResumeAllAction::Unknown,
            format!(
                "Session is not running but no terminal result could be resolved: {}",
                if session_name.is_empty() {
                    record.uuid.clone()
                } else {
                    session_name.clone()
                }
            ),
            None,
        ));
    }

    ExecutionResumeAllResult {
        success: true,
        output: Some(format_resume_all(&entries, output_format)),
        error: None,
    }
}

#[cfg(test)]
#[path = "execution_resume_all_cases.rs"]
mod tests;
